lib/chef/resource/openssl_dhparam.rb

```diff
@@ -58,18 +58,16 @@ class OpensslDhparam < Chef::Resource
       action :create do
         description "Create the dhparam file."
 
-        unless dhparam_pem_valid?(new_resource.path)
-          converge_by("Create a dhparam file #{new_resource.path}") do
-            dhparam_content = gen_dhparam(new_resource.key_length, new_resource.generator).to_pem
+        converge_by("Create a dhparam file #{new_resource.path}") do
```

Comment:
This will converge on each run.

Reply:
This `converge_by` is just a message. File is not being created if already exist. But `mode` property can be updated. And also with this fix content of the existing file is also not changing on subsequent run. Ensured idempotency of the resource.

## Recipe
```
root@dh-openssl-param:~/chef# cat ../openssl_dhparam.rb
openssl_dhparam '/etc/ssl/certs/dhparam.pem' do
    mode '0640'
end
```
## First Run
```
root@dh-openssl-param:~/chef# bundle exec chef-apply ../openssl_dhparam.rb
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * openssl_dhparam[/etc/ssl/certs/dhparam.pem] action create
    - Create a dhparam file /etc/ssl/certs/dhparam.pem
    * file[/etc/ssl/certs/dhparam.pem] action create
      - create new file /etc/ssl/certs/dhparam.pem
      - update content in file /etc/ssl/certs/dhparam.pem from none to 302bc8
      - suppressed sensitive resource
      - change mode from '' to '0640'
root@dh-openssl-param:~/chef#
```
## Second Run
```
root@dh-openssl-param:~/chef# bundle exec chef-apply ../openssl_dhparam.rb
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * openssl_dhparam[/etc/ssl/certs/dhparam.pem] action create
    - Create a dhparam file /etc/ssl/certs/dhparam.pem
    * file[/etc/ssl/certs/dhparam.pem] action create (up to date)

root@dh-openssl-param:~/chef#
```
## If `mode` is changed
```
root@dh-openssl-param:~/chef# bundle exec chef-apply ../openssl_dhparam.rb
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * openssl_dhparam[/etc/ssl/certs/dhparam.pem] action create
    - Create a dhparam file /etc/ssl/certs/dhparam.pem
    * file[/etc/ssl/certs/dhparam.pem] action create
      - change mode from '0640' to '0600'

root@dh-openssl-param:~/chef#
```
@tas50 Please let me know your thoughts.

Reply:
converge_by is not just a log message though. converge_by marks the resource as being updated, which fires off notifications and adds it to any resource reporting which goes to A2. It's pretty important that we only do this when a resource is actually changed. If not bad things happen.